### DISCUSSION TITLE
Archived Q&A

### DISCUSSION THREAD

**Discussion by ruudvdent** [WAM2layers/WAM2layers#186] [Q&A] [answered]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by WenyuZhou-4** [comment]
Hello, I just begin to try the WAM2layers. I've downloeaded the example data 'floodcase_202107.yaml' as well as the preprocesseed data file from the google derive.
When I fellow the document and run `wam2layers preprocess era5 floodcase_202107.yaml`, I get errors like:

  File "/home/wenyuzhou/miniconda3/envs/wamenv/lib/pypy3.9/site-packages/pydantic/main.py", line 341, in __init__
    raise validation_error
pydantic.error_wrappers.ValidationError: 6 validation errors for Config
filename_template
  field required (type=value_error.missing)
input_frequency
  field required (type=value_error.missing)
output_frequency
  field required (type=value_error.missing)
level_type
  field required (type=value_error.missing)
levels
  field required (type=value_error.missing)
log_level
  field required (type=value_error.missing)

The only thing I changed in the `floodcase_202107.yaml` file is the `input_folder`. Can you tell me what's wrong with it?

**Reply by Peter9192** [reply]
Hello Wenyu, welcome to the discussion forum! 

I think the example data with example config file are slightly outdated with respect to the latest version of the code. Please have a look at the [documentation](https://wam2layers.readthedocs.io/en/latest/config.html) on the configuration file, and the latest example config file [here](https://github.com/WAM2layers/WAM2layers/blob/main/example_config.yaml). Make sure that you understand all the settings and that they are set correctly for your case, for example, the paths should reflect the paths on your system. 

Let us know if you still have questions.

**Reply by WenyuZhou-4** [reply]
In the latest version there seems an issue entering the dates:

In the config I have:
# Dates
preprocess_start_date: "2021-07-01"
preprocess_end_date: "2021-07-15"
track_start_date: "2021-07-01"
track_end_date: "2021-07-15"
event_start_date: "2021-07-13"
event_end_date: "2021-07-15"

Which yields the following errors:

ValidationError: 6 validation errors for Config
preprocess_start_date
  invalid datetime format (type=value_error.datetime)
preprocess_end_date
  invalid datetime format (type=value_error.datetime)
track_start_date
  invalid datetime format (type=value_error.datetime)
track_end_date
  invalid datetime format (type=value_error.datetime)
event_start_date
  invalid datetime format (type=value_error.datetime)
event_end_date
  invalid datetime format (type=value_error.datetime)

Anyone else had the same problem? Does this mean the example should be updated?

**Reply by Peter9192** [reply]
Ahh, it looks like you're dates are missing a timestamp. That's been [changed somewhat](https://github.com/WAM2layers/WAM2layers/pull/176) recently. Can you try adding times like so:
```py
preprocess_start_date: "2021-07-01T00:00"
```
This has been documented in https://wam2layers.readthedocs.io/en/latest/config.html#wam2layers.config.Config.preprocess_start_date but apparently, we've forgotten to update the example config.

**Reply by Peter9192** [reply]
Made a new issue for updating the example config file. https://github.com/WAM2layers/WAM2layers/issues/189

**Reply by Peter9192** [reply]
#194 fixes the datetime issue in the example config file

**Comment by PoornimaUnnikrishnan** [comment]
Hello, I am in the initial phase of exploring WAM2layers in tracking the moisture source of a region in California.
 
•              Data used: ERA 5 for California 
 
•              While backtracking using WAM2 layers, the e_track component is found to be zero and hence cannot plot the contour levels.
 
•              While checking for the source of this error, it is found that while pre-processing the data, tracked moisture content (s_upper and s_lower) became zero. I am trying to figure out why this error is occurring repeatedly. Can you please suggest strategies that can be used to diagnose such errors and correct these issues? Thanks.

**Reply by Peter9192** [reply]
Hi @PoornimaUnnikrishnan welcome to the discussion forum! 
First some clarification, `s_upper` and `s_lower` in the pre-processing are not the _tracked_ moisture but rather the _total_ moisture in the respective layers. Regardless, you're right that they shouldn't become zero. It's hard to pinpoint exactly what's causing your issue, so I'm going to recommend a more general approach to figure out what's going on. This requires that you install WAM2layers in development model, i.e. with a local, editable installation of the source code. See the [instructions for developers here](https://wam2layers.readthedocs.io/en/latest/develop.html). Then, what I like to do is put a debug statement somewhere in the code like so:

```py
import IPython; IPython.embed(); quit()
```

When you run the code again, as soon as it reaches this line, it will start a new interactive IPython session, and you can inspect all the variables at that point in the code. By moving this debug statement around in the [preprocessing file](https://github.com/WAM2layers/WAM2layers/blob/main/src/wam2layers/preprocessing/era5.py), you should be able to pinpoint where `s_upper` and `s_lower` becomes 0. 

I wish I could be more helpful. Good luck, and don't hesitate to reach out again.

**Reply by PoornimaUnnikrishnan** [reply]
Thanks for the reply. I couldn't figure out why s_upper and s_lower becomes zero while I am using model_levels as lev_type.
When I'm using the level_type as pressure levels, at some dates in input data, assertion error occurs stating pressure levels should increase monotonically.

**Reply by Peter9192** [reply]
So the data that you're using - is it on pressure levels or on model levels?

**Reply by rampting** [reply]
Hi @Peter9192 and @PoornimaUnnikrishnan,

I may have found a related problem that has to do with s_lower and s_upper after the preprocessing step. Please let me share with you my graphs, where I have visualized some nedcdf files with Panoply. 

![s_upper_lower error after preprocessing](https://user-images.githubusercontent.com/124769652/226943690-37856981-d239-4f3d-8ede-62d0f60f1e56.png)

- I checked my input data which all seem fine. As an example I included specific humidity. 
- Furthermore, after the preprocessing step my fluxes also seem correct (fx_upper) for my domain. However, something is off about the s_upper and s_lower.
- As you can see, they do not visualize my complete domain. s_upper only gives values for the upper area of my domain, with all other grids "NAN". s_lower only gives values for the rest, except for the s_upper area. In this case s_lower gives "NAN" for this specific part. 

Would anyone be able to address the issue? I will try and find something in the preprocesssing code that could be off, however I am not an expert in this field of coding. 

Thank you in advance,

Best, Rens

**Reply by ruudvdent** [reply]
Thanks Rens for posting. Seems like possibly a bug slipped into the software in the recent updates, or at least this requires some investigation. Could you check if this s_upper and s_lower issue also occurs for the example floodcase data?

**Reply by rampting** [reply]
If I return to the old way of setting s_upper and s_lower in the era5 file for preprocessing, the domains seem correct: 

![image](https://user-images.githubusercontent.com/124769652/227572307-dd1a8092-538e-43c8-8d9d-442b9151da28.png)

![image](https://user-images.githubusercontent.com/124769652/227572487-132cc92e-8d58-4e94-8e52-7ab0dbdbf2d8.png)

However, now my backtracking is not working. Previously, the backtracking is not working even though my preprocessed file seems right.

**Reply by vincentdefeiter** [reply]
Yes, switching it back to old s_upper and s_lower changed it. However, there the backtracking with the files of @rampting do not work. The files of the flood case and the pre-processing files for my project seem to work.

**Reply by rampting** [reply]
Indeed. Backtrack for the floodcase202107_example does work. However the time length of this netcdf file is 25.

My time length after preprocessing is 24. Could this be a problem?

**Reply by rampting** [reply, marked as answer]
@Peter9192 @ruudvdent @vincentdefeiter 

I am quite sure the issue is resolved. It has to do with the setup of your configuration file.

1. I changed preprocessing file "era5" to the old setup of defining s_lower and s_upper. Green is "the old setup", so what I just changed. Red is the "new setup" which is not working.

![image](https://user-images.githubusercontent.com/124769652/228008387-f8352374-c19a-4954-8389-78c61d9ab064.png)

2. change the yaml configuration file:

The processing and backtracking step seem require different settings for how you define your model levels. 

for preprocess you should specifically define "level_type", and "levels". Whilst backtrack uses "modellevels" as input. Also I included the following five lines (and keeping the rest, otherwise preprocessing does not work anymore):

- modellevels: [20,40,60,80,90,95,100,105,110,115,120,123,125,128,130,131,132,133,134,135,136,137]
- vertical_integral_available: True
- timetracking: false
- distancetracking: false
- name_of_run: 'default_run'

I am not sure if the above 4 lines helped in resolving the issue. but mine works now

![image](https://user-images.githubusercontent.com/124769652/228009236-68274479-a558-4ea3-901a-18ec82e4557f.png)
![image](https://user-images.githubusercontent.com/124769652/228009296-31819953-52a6-4562-be2c-24d9aeb4a4d4.png)

**Reply by Peter9192** [reply]
This change between older/newer selection in s_upper rings a bell. See https://github.com/WAM2layers/WAM2layers/commit/457b29d4a8b8fdb5d7004575c9fa4d1a191554c6. 

I should find some time to have a better look at this. It's great to see that you are all digging into the code and finding issues. This will surely help to improve the model in the long run!

**Reply by YiyingWang-00** [reply]
@rampting Thanks for your solution, I have got the right preprocessing data. But in backtrack, the value in e_track and tagged_precipitation is all 0. Have you met this problem?I don't know how to solve it. I sent a new question in the discussion.
![微信图片_20230419232013](https://user-images.githubusercontent.com/112865233/233128802-1d7dab70-a653-47a2-8851-92c1b28a796e.png)
![微信图片_20230419232532](https://user-images.githubusercontent.com/112865233/233128821-555d0252-3007-472e-9ff3-8e3938d53ae3.png)
@vincentdefeiter

**Comment by PoornimaUnnikrishnan** [comment]
Hi,
As I previously mentioned there are problems with s_lower and s_upper when I am using model_levels as level type. 
 
![s_lower_in_2017-01-07_fluxes_storages](https://user-images.githubusercontent.com/128003598/227587001-70890c3c-164f-414f-acd4-14e252823d5d.png)

Fig1.
 


![fx_lower_in_2017-05-29_fluxes_storages](https://user-images.githubusercontent.com/128003598/227587078-8a2c65b3-80bd-4cc1-a640-90341ff78ab9.png)
Fig2.
Hence, I tried using era5 on pressure levels and tried preprocessing, at some dates in input data, assertion error occurs stating pressure levels should increase monotonically. Is this due to error in the era5 data? I tried it for different data sets (ERA5 for different time period) the issue still persists.

With the preprocessed data that I could get before the assertion error, I carried out the backtracking, however the output files are not correctly georeferenced, and do not contain units. The e_track values doesn't seem right without the unit (see fig 4, which is the output from  the visualization in wam2layers.

 
![e_track_in_2018-02-11_s_track](https://user-images.githubusercontent.com/128003598/227587191-e8053858-2041-4dad-b971-59c5c7010abe.png)

Fig 3 
![cumulative_sources 2018](https://user-images.githubusercontent.com/128003598/227587234-86ee744f-3c0a-4ef3-b960-6269b41aea67.png)

Fig4.

**Comment by Peter9192** [comment]
I'm afraid this discussion is not focussed enough and we'll loose track of all the issues discussed here. Shall we change the topic of this discussion, and try to have one discussion per issue/question/topic?

**Reply by ruudvdent** [reply]
Yes, makes sense! If you have a good suggestion of how to split, please go ahead.
